tools/functional-tester/etcd-runner/main.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2016 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"math/rand"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"golang.org/x/net/context"
+)
+
+func main() {
+	log.SetFlags(log.Lmicroseconds)
+
+	endpointStr := flag.String("endpoints", "localhost:2379", "endpoints of etcd cluster")
+	mode := flag.String("mode", "lock-racer", "test mode (lock-racer)")
+	round := flag.Int("rounds", 100, "number of rounds to run")
+	flag.Parse()
+	eps := strings.Split(*endpointStr, ",")
+
+	switch *mode {
+	case "lock-racer":
+		runRacer(eps, *round)
+	case "lease-renewer":
+		runLeaseRenewer(eps)
+	default:
+		fmt.Fprintf(os.Stderr, "unsupported mode %v\n", *mode)
+	}
+}
+
+func runLeaseRenewer(eps []string) {
+	c := randClient(eps)
+	ctx := context.Background()
+
+	for {
+		var (
+			l   *clientv3.LeaseGrantResponse
+			lk  *clientv3.LeaseKeepAliveResponse
+			err error
+		)
+		for {
+			l, err = c.Lease.Grant(ctx, 5)
+			if err == nil {
+				fmt.Printf("grant lease %d\n", l.ID)
+				break
+			}
+			fmt.Println("grant err", err)
+		}
+		expire := time.Now().Add(time.Duration(l.TTL-1) * time.Second)
+
+		for {
+			log.Printf("about to renew lease")
+			lk, err = c.Lease.KeepAliveOnce(ctx, l.ID)
+			if grpc.Code(err) == codes.NotFound {
+				if time.Now().Sub(expire) < 0 {
+					log.Printf("bad renew! exceeded: %v", time.Now().Sub(expire))
+					for {
+						lk, err = c.Lease.KeepAliveOnce(ctx, l.ID)
+						fmt.Println(lk, err)
+						time.Sleep(time.Second)
+					}
+				}
+				log.Printf("lost lease %d, expire: %v\n", l.ID, expire)
+				break
+			}
+			if err != nil {
+				fmt.Println("renew err", err)
+				continue
+			}
+			expire = time.Now().Add(time.Duration(lk.TTL-1) * time.Second)
+			log.Printf("renewed lease %d, expire: %v\n", lk.ID, expire)
+			time.Sleep(time.Duration(lk.TTL-2) * time.Second)
+		}
+	}
+}
+
+func runRacer(eps []string, round int) {
+	nrace := 15
+	prefix := "racers"
+	racers := make([]*concurrency.Mutex, nrace)
+	progress := make([]int, nrace)
+	finished := make(chan struct{}, 0)
+
+	var (
+		mu  sync.Mutex
+		cnt int
+	)
+	ctx := context.Background()
+
+	var wg sync.WaitGroup
+
+	for i := range racers {
+		racers[i] = concurrency.NewMutex(randClient(eps), prefix)
```

Comment:
yea... but this test should be a long running one and should "never" end... so i am less worry about it.